### PR TITLE
feat!: Add MaxCollateralShare, MaxSupplyUtilization, and MinCollateralLiquidity

### PR DESCRIPTION
**Pull request by toteki**
## Description

- Adds new params to proto

Also
- Implements `MaxSupplyUtilization`
- Does not implement the other new params

closes: #1093 #1095 #1036 and some other things that weren't made into issues first

Currently backtracks on #1107 due to second thoughts - see discussion

---

### Author Checklist

_All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues._

I have...

- [x] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [x] added `!` to the type prefix if API or client breaking change
- [x] added appropriate labels to the PR
- [x] targeted the correct branch (see [PR Targeting](https://github.com/umee-network/umee/blob/main/CONTRIBUTING.md#pr-targeting))
- [x] provided a link to the relevant issue or specification
- [x] added a changelog entry to `CHANGELOG.md`
- [x] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [x] updated the relevant documentation or specification
- [ ] reviewed "Files changed" and left comments if necessary
- [ ] confirmed all CI checks have passed

### Reviewers Checklist

_All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items._

I have...

- [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed all author checklist items have been addressed
- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -67,6 +67,7 @@ Ref: https://keepachangelog.com/en/1.0.0/
 - [1096](https://github.com/umee-network/umee/pull/1096) Add `max_collateral_share` to the x/leverage token registry.
 - [1094](https://github.com/umee-network/umee/pull/1094) Added TotalCollateral query.
 - [1099](https://github.com/umee-network/umee/pull/1099) Added TotalBorrowed query.
+- [11XX](https://github.com/umee-network/umee/pull/11XX) Add `max_borrow_utilization` and `min_collateral_liquidity` to the x/leverage token registry.
```
Comment (toteki) [inline CHANGELOG.md:70]:
```suggestion
- [1141](https://github.com/umee-network/umee/pull/1141) Add `max_borrow_utilization` and `min_collateral_liquidity` to the x/leverage token registry.
```

File: docs/architecture/ADR-010-market-params.md [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+### Maximum Collateral Share
+
+Another proposed parameter is `MaxCollateralShare`, to be defined per token.
+
+```go
+total_collateral_value(token) = total_collateral(token) * oracle_price_usd(token)
+collateral_share(token) = total_collateral_value(token) / Sum_All_Tokens(total_collateral_value) // ranges 0 - 1
+```
+
+Implementing `MaxCollateralShare` would restrict `MsgAddCollateral` from increasing `total_collateral` above a desired level.
```
Comment (toteki) [inline docs/architecture/ADR-010-market-params.md:83]:
```suggestion
Implementing `MaxCollateralShare` would restrict `MsgCollateralize` from increasing `total_collateral` above a desired level.
```

File: docs/architecture/ADR-010-market-params.md [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+
+High collateral utilization is dangerous for the system: When collateral utilization is above 1, liquidators may not be able to withdraw their the liquidated collateral.
+
+Let's draw the following scenario to picture the liquidators risk:
+
+> | - | Supply | Collateral | Borrowed |
+> | Alice| $1.2M BTC | - | - |
+> | Bob | - | $1.5M LUNA | $1M BTC |
+> | Charlie | - | $2M BTC | $1.4M LUNA |
+>
```
Comment (toteki) [inline docs/architecture/ADR-010-market-params.md:139]:
```suggestion
| - | Supply | Collateral | Borrowed |
| Alice| $1.2M BTC | - | - |
| Bob | - | $1.5M LUNA | $1M BTC |
| Charlie | - | $2M BTC | $1.4M LUNA |

```

File: docs/architecture/ADR-010-market-params.md [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+
+High collateral utilization is dangerous for the system: When collateral utilization is above 1, liquidators may not be able to withdraw their the liquidated collateral.
+
+Let's draw the following scenario to picture the liquidators risk:
+
+| - | Supply | Collateral | Borrowed |
+| Alice| $1.2M BTC | - | - |
+| Bob | - | $1.5M LUNA | $1M BTC |
+| Charlie | - | $2M BTC | $1.4M LUNA |
+
```
Comment (toteki) [inline docs/architecture/ADR-010-market-params.md:139]:
```suggestion
> | User | Supply | Collateral | Borrowed |
> | - | - | - | - |
> | Alice| $1.2M BTC | - | - |
> | Bob | - | $1.5M LUNA | $1M BTC |
> | Charlie | - | $2M BTC | $1.4M LUNA |
>
```

File: docs/architecture/ADR-010-market-params.md [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+```
+
+### Maximum Supply Utilization
+
+One proposed parameter is `MaxSupplyUtilization`, to be defined per token.
+
+```go
+supply_utilization(token) = total_borrowed(token) / total_supply(token) // ranges 0 - 1
+```
+
```
Comment (toteki) [inline docs/architecture/ADR-010-market-params.md:55]:
```suggestion

For example, if `1000 ATOM` are supplied and `200 ATOM` are borrowed, `supply_utilization` is `0.2`.

```

File: docs/architecture/ADR-010-market-params.md [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+### Maximum Collateral Share
+
+Another proposed parameter is `MaxCollateralShare`, to be defined per token.
+
+```go
+total_collateral_value(token) = total_collateral(token) * oracle_price_usd(token)
+collateral_share(token) = total_collateral_value(token) / Sum_All_Tokens(total_collateral_value) // ranges 0 - 1
+```
+
```
Comment (toteki) [inline docs/architecture/ADR-010-market-params.md:82]:
```suggestion

For example, if `10 ATOM` collateral exists at a price of `$20` alongside `100 OSMO` collateral at a price of `$1`, then `collateral_share(ATOM)` is `10 * $20 / $300 = 0.66` and `collateral_share(OSMO)` is `$100 / $300 = 0.33`.

```

File: docs/architecture/ADR-010-market-params.md [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+### Minimum Collateral Liquidity
+
+Another proposed parameter is `MinCollateralLiquidity`, to be defined per token.
+
+It is the reciprocal of `MaxCollateralUtilization`, functionally equivalent in which messages it would restrict.
+
+```go
+collateral_liquidity(token) = available(token) / total_collateral(token) // ranges 0 - ∞
+```
+
```
Comment (toteki) [inline docs/architecture/ADR-010-market-params.md:96]:
```suggestion

For example if `200 ATOM` of collateral exists and the `x/leverage` module only has `160 ATOM` available (with the rest of supply being borrowed out), then `collateral_liquidity(ATOM) = 0.8`.

```

File: docs/architecture/ADR-010-market-params.md [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+### Collateral Utilization
+
+One proposed parameter is `MaxCollateralUtilization`, to be defined per token.
+
+It is the reciprocal of `MinCollateralLiquidity`, so the motivation and tradeoffs will be the same.
+
+```go
+collateral_utilization(token) = total_collateral(token) / available(token) // ranges 0 - ∞
+```
+
```
Comment (toteki) [inline docs/architecture/ADR-010-market-params.md:124]:
```suggestion

For example if `200 ATOM` of collateral exists and the `x/leverage` module only has `40 ATOM` available (with the rest of supply being borrowed out), then `collateral_utilization(ATOM) = 5`.

```

**Comment by toteki** [comment]
This PR as been split up:
- Documentation (#1152)
- Proto (#1159)
After those merge, I will take the remaining code in this PR (i.e. the implementation of MaxSupplyUtilization and related calculations), turn them into a small PR, and close this.

**Comment by robert-zaremba** [comment]
seams like we can close now to avoid confusions. You can keep your branch locally.

**Pull request closed without merging**